src/lib.rs

```diff
@@ -172,34 +169,93 @@ impl<R: Read> Archive<R> {
                                           "empty file name in tarball"))
             }
             let is_directory = bytes[bytes.len() - 1] == b'/';
+
+            // Notes regarding bsdtar 2.8.3 / libarchive 2.8.3:
+            // * Leading '/'s are trimmed. For example, `///test` is treated as
+            //   `test`.
+            // * If the filename contains '..', then the file is skipped when
+            //   extracting the tarball.
+            // * '//' within a filename is effectively skipped. An error is
+            //   logged, but otherwise the effect is as if any two or more
+            //   adjacent '/'s within the filename were consolidated into one
+            //   '/'.
+
             let mut dst = into.to_path_buf();
+            let mut contains_dot_dot = false;
+            let mut seen_non_empty_part = false;
             for part in bytes.split(|x| *x == b'/') {
+                if part == b".." {
+                    contains_dot_dot = true;
+                    break;
```

Comment:
I think this would be better expressed by removing the `contains_dot_dot` variable and instead labeling the outer loop with `'outer:` and then here saying `break 'outer`